File: forge-plugin/skills/generate-jest-unit-tests/scripts/test_analyzer.ts
#!/usr/bin/env ts-node
/**
 * Angular Test Analyzer Helper Script
 *
 * This script helps analyze existing Angular test files to understand:
 * - Testing framework used (Jest vs Jasmine/Karma)
 * - Test file naming conventions
 * - TestBed configuration patterns
 * - Mock creation patterns
 * - Common test structures
 *
 * Usage:
 *     ts-node test_analyzer.ts <test_directory>
 *     ts-node test_analyzer.ts <test_file.spec.ts>
 */

import * as fs from 'fs';
import * as path from 'path';

interface TestAnalysis {
  file: string;
  framework: 'jest' | 'jasmine' | 'unknown';
  testCount: number;
  usesTestBed: boolean;
  usesHttpMocking: boolean;
  usesMockStore: boolean;
  describeBlocks: string[];
  mockPatterns: string[];
  angularVersion?: string;
}

interface AggregateAnalysis {
  testFilesCount: number;
  totalTests: number;
  primaryFramework: 'jest' | 'jasmine' | 'mixed';
  frameworkCounts: { jest: number; jasmine: number };
  testBedUsage: string;
  httpMockingPattern: string;
  ngrxUsage: string;
  testFiles: string[];
  namingPattern: string;
  commonPatterns: string[];
}

function analyzeTestFile(filePath: string): TestAnalysis {
  const content = fs.readFileSync(filePath, 'utf-8');

  return {
    file: filePath,
    framework: detectFramework(content),
    testCount: countTests(content),
    usesTestBed: content.includes('TestBed.configureTestingModule'),
    usesHttpMocking: detectHttpMocking(content),
    usesMockStore: content.includes('provideMockStore') || content.includes('MockStore'),
    describeBlocks: findDescribeBlocks(content),
    mockPatterns: findMockPatterns(content),
    angularVersion: detectAngularVersion(content),
  };
}

function detectFramework(content: string): 'jest' | 'jasmine' | 'unknown' {
  const hasJest =
    content.includes('jest.fn()') ||
    content.includes('jest.mock') ||
    content.includes('jest.spyOn') ||
    content.includes('mockReturnValue');

  const hasJasmine =
    content.includes('jasmine.createSpyObj') ||
    content.includes('spyOn(') ||
    content.includes('.and.returnValue') ||
    content.includes('HttpTestingController');

  if (hasJest && !hasJasmine) return 'jest';
  if (hasJasmine && !hasJest) return 'jasmine';
  if (hasJest && hasJasmine) return 'jest'; // Assume migration to Jest
  return 'unknown';
}

function countTests(content: string): number {
  const itMatches = content.match(/\bit\s*\(/g);
  return itMatches ? itMatches.length : 0;
}

function detectHttpMocking(content: string): boolean {
  return (
    content.includes('HttpTestingController') ||
    content.includes('HttpClient') ||
    content.includes('httpMock')
  );
}

function findDescribeBlocks(content: string): string[] {
  const describeRegex = /describe\s*\(\s*['"`]([^'"`]+)['"`]/g;
  const matches: string[] = [];
  let match;

  while ((match = describeRegex.exec(content)) !== null) {
    matches.push(match[1]);
  }

  return matches;
}

function findMockPatterns(content: string): string[] {
  const patterns: string[] = [];

  if (content.includes('jasmine.createSpyObj')) {
    patterns.push('jasmine.createSpyObj');
  }
  if (content.includes('jest.fn()')) {
    patterns.push('jest.fn()');
  }
  if (content.includes('jest.mock(')) {
    patterns.push('jest.mock()');
  }
  if (content.includes('provideMockStore')) {
    patterns.push('provideMockStore (NgRx)');
  }
  if (content.includes('MockComponent')) {
    patterns.push('MockComponent');
  }
  if (content.includes('NO_ERRORS_SCHEMA')) {
    patterns.push('NO_ERRORS_SCHEMA');
  }

  return patterns;
}

function detectAngularVersion(content: string): string | undefined {
  // Try to detect standalone components (Angular 14+)
  if (content.includes('standalone: true')) {
    return '14+';
  }
  // Try to detect signals (Angular 16+)
  if (content.includes('signal(') || content.includes('computed(')) {
    return '16+';
  }
  return undefined;
}

function findTestFiles(directory: string): string[] {
  if (!fs.existsSync(directory)) {
    return [];
  }

  const stat = fs.statSync(directory);
  if (stat.isFile()) {
    return isTestFile(directory) ? [directory] : [];
  }

  const files: string[] = [];
  const entries = fs.readdirSync(directory, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(directory, entry.name);
    if (entry.isDirectory()) {
      files.push(...findTestFiles(fullPath));
    } else if (entry.isFile() && isTestFile(fullPath)) {
      files.push(fullPath);
    }
  }

  return files;
}

function isTestFile(filePath: string): boolean {
  return filePath.endsWith('.spec.ts') || filePath.endsWith('.test.ts');
}

function analyzeDirectory(directory: string): AggregateAnalysis | { error: string } {
  const testFiles = findTestFiles(directory);

  if (testFiles.length === 0) {
    return {
      error: `No test files found in ${directory}`,
    };
  }

  const analyses = testFiles.map(analyzeTestFile);

  const frameworkCounts = {
    jest: analyses.filter(a => a.framework === 'jest').length,
    jasmine: analyses.filter(a => a.framework === 'jasmine').length,
  };

  const totalTests = analyses.reduce((sum, a) => sum + a.testCount, 0);
  const testBedCount = analyses.filter(a => a.usesTestBed).length;
  const httpMockingCount = analyses.filter(a => a.usesHttpMocking).length;
  const ngrxCount = analyses.filter(a => a.usesMockStore).length;

  let primaryFramework: 'jest' | 'jasmine' | 'mixed';
  if (frameworkCounts.jest > frameworkCounts.jasmine) {
    primaryFramework = 'jest';
  } else if (frameworkCounts.jasmine > frameworkCounts.jest) {
    primaryFramework = 'jasmine';
  } else {
    primaryFramework = 'mixed';
  }

  const commonPatterns: string[] = [];
  const allMockPatterns = analyses.flatMap(a => a.mockPatterns);
  const patternCounts = new Map<string, number>();
  allMockPatterns.forEach(pattern => {
    patternCounts.set(pattern, (patternCounts.get(pattern) || 0) + 1);
  });
  patternCounts.forEach((count, pattern) => {
    if (count >= testFiles.length * 0.3) { // Used in 30%+ of files
      commonPatterns.push(`${pattern} (${count} files)`);
    }
  });

  return {
    testFilesCount: testFiles.length,
    totalTests,
    primaryFramework,
    frameworkCounts,
    testBedUsage: `${testBedCount}/${testFiles.length} files`,
    httpMockingPattern: `${httpMockingCount}/${testFiles.length} files`,
    ngrxUsage: `${ngrxCount}/${testFiles.length} files`,
    testFiles: testFiles.slice(0, 10), // First 10 files
    namingPattern: detectNamingPattern(testFiles),
    commonPatterns,
  };
}

function detectNamingPattern(testFiles: string[]): string {
  const specCount = testFiles.filter(f => f.endsWith('.spec.ts')).length;
  const testCount = testFiles.filter(f => f.endsWith('.test.ts')).length;

  if (specCount > testCount) {
    return '*.spec.ts (Angular standard)';
  } else if (testCount > specCount) {
    return '*.test.ts';
  } else {
    return 'mixed';
  }
}

function printAnalysis(analysis: AggregateAnalysis | { error: string }): void {
  if ('error' in analysis) {
    console.log(`Error: ${analysis.error}`);
    console.log('Searched for: *.spec.ts, *.test.ts files');
    return;
  }

  console.log('='.repeat(70));
  console.log('ANGULAR TEST ANALYSIS');
  console.log('='.repeat(70));
  console.log(`\nTest Files Found: ${analysis.testFilesCount}`);
  console.log(`Total Tests: ${analysis.totalTests}`);
  console.log(`Primary Framework: ${analysis.primaryFramework}`);
  console.log(`  - Jest: ${analysis.frameworkCounts.jest} files`);
  console.log(`  - Jasmine: ${analysis.frameworkCounts.jasmine} files`);
  console.log(`Naming Pattern: ${analysis.namingPattern}`);
  console.log(`\nTestBed Usage: ${analysis.testBedUsage}`);
  console.log(`HTTP Mocking: ${analysis.httpMockingPattern}`);
  console.log(`NgRx (MockStore): ${analysis.ngrxUsage}`);

  if (analysis.commonPatterns.length > 0) {
    console.log(`\nCommon Mock Patterns:`);
    analysis.commonPatterns.forEach(pattern => {
      console.log(`  - ${pattern}`);
    });
  }

  console.log(`\nSample Test Files:`);
  analysis.testFiles.forEach(file => {
    console.log(`  - ${file}`);
  });

  if (analysis.testFilesCount > 10) {
    console.log(`  ... and ${analysis.testFilesCount - 10} more`);
  }

  console.log('='.repeat(70));
}

function printFileAnalysis(analysis: TestAnalysis): void {
  console.log('='.repeat(70));
  console.log(`ANALYSIS: ${analysis.file}`);
  console.log('='.repeat(70));
  console.log(`Framework: ${analysis.framework}`);
  console.log(`Test Count: ${analysis.testCount}`);
  console.log(`Uses TestBed: ${analysis.usesTestBed}`);
  console.log(`Uses HTTP Mocking: ${analysis.usesHttpMocking}`);
  console.log(`Uses NgRx MockStore: ${analysis.usesMockStore}`);

  if (analysis.angularVersion) {
    console.log(`Angular Version: ${analysis.angularVersion}`);
  }

  if (analysis.describeBlocks.length > 0) {
    console.log(`\nDescribe Blocks:`);
    analysis.describeBlocks.forEach(block => {
      console.log(`  - ${block}`);
    });
  }

  if (analysis.mockPatterns.length > 0) {
    console.log(`\nMock Patterns:`);
    analysis.mockPatterns.forEach(pattern => {
      console.log(`  - ${pattern}`);
    });
  }

  console.log('='.repeat(70));
}

function main(): void {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.log('Usage: ts-node test_analyzer.ts <test_directory_or_file>');
    console.log('\nExample:');
    console.log('  ts-node test_analyzer.ts src/app');
    console.log('  ts-node test_analyzer.ts src/app/components/user.component.spec.ts');
    process.exit(1);
  }

  const target = args[0];

  if (!fs.existsSync(target)) {
    console.error(`Error: Path '${target}' does not exist`);
    process.exit(1);
  }

  const stat = fs.statSync(target);

  if (stat.isFile()) {
    // Analyze single file
    const analysis = analyzeTestFile(target);
    printFileAnalysis(analysis);
  } else {
    // Analyze directory
    const analysis = analyzeDirectory(target);
    printAnalysis(analysis);
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}

export {
  analyzeTestFile,
  analyzeDirectory,
  TestAnalysis,
  AggregateAnalysis,
};

File: forge-plugin/skills/generate-python-unit-tests/scripts/test_analyzer.py
#!/usr/bin/env python3
"""
Test Analyzer Helper Script

This script helps analyze existing Python test files to understand:
- Testing framework used (pytest, unittest, etc.)
- Test file naming conventions
- Fixture patterns
- Common test structures

Usage:
    python test_analyzer.py <test_directory>
    python test_analyzer.py <test_file.py>
"""

import os
import re
import sys
from pathlib import Path
from collections import defaultdict


def analyze_test_file(file_path):
    """
    Analyze a single test file and extract patterns.

    Args:
        file_path: Path to the test file

    Returns:
        dict: Analysis results
    """
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    analysis = {
        'file': str(file_path),
        'framework': detect_framework(content),
        'fixtures': find_fixtures(content),
        'test_count': count_tests(content),
        'uses_mocking': 'mock' in content.lower() or 'patch' in content,
        'uses_parametrize': '@pytest.mark.parametrize' in content,
        'test_classes': find_test_classes(content),
        'imports': extract_imports(content),
    }

    return analysis


def detect_framework(content):
    """Detect which testing framework is used."""
    if 'import pytest' in content or 'from pytest' in content:
        return 'pytest'
    elif 'import unittest' in content or 'from unittest' in content:
        return 'unittest'
    elif 'import nose' in content:
        return 'nose'
    else:
        return 'unknown'


def find_fixtures(content):
    """Find pytest fixtures in the file."""
    fixture_pattern = r'@pytest\.fixture(?:\([^)]*\))?\s+def\s+(\w+)'
    fixtures = re.findall(fixture_pattern, content)
    return fixtures


def count_tests(content):
    """Count the number of test functions/methods."""
    # Match both standalone functions and methods
    test_pattern = r'def\s+(test_\w+)'
    tests = re.findall(test_pattern, content)
    return len(tests)


def find_test_classes(content):
    """Find test classes."""
    class_pattern = r'class\s+(Test\w+)'
    classes = re.findall(class_pattern, content)
    return classes


def extract_imports(content):
    """Extract import statements."""
    import_lines = [
        line.strip()
        for line in content.split('\n')
        if line.strip().startswith(('import ', 'from '))
    ]
    return import_lines


def analyze_directory(directory):
    """
    Analyze all test files in a directory.

    Args:
        directory: Path to directory containing tests

    Returns:
        dict: Aggregate analysis results
    """
    test_files = find_test_files(directory)

    if not test_files:
        return {
            'error': f'No test files found in {directory}',
            'searched_patterns': ['test_*.py', '*_test.py']
        }

    analyses = [analyze_test_file(f) for f in test_files]

    # Aggregate results
    frameworks = defaultdict(int)
    all_fixtures = []
    total_tests = 0
    uses_mocking = 0
    uses_parametrize = 0

    for analysis in analyses:
        frameworks[analysis['framework']] += 1
        all_fixtures.extend(analysis['fixtures'])
        total_tests += analysis['test_count']
        if analysis['uses_mocking']:
            uses_mocking += 1
        if analysis['uses_parametrize']:
            uses_parametrize += 1

    return {
        'test_files_count': len(test_files),
        'total_tests': total_tests,
        'primary_framework': max(frameworks.items(), key=lambda x: x[1])[0] if frameworks else 'none',
        'frameworks': dict(frameworks),
        'common_fixtures': list(set(all_fixtures)),
        'mocking_usage': f'{uses_mocking}/{len(test_files)} files',
        'parametrize_usage': f'{uses_parametrize}/{len(test_files)} files',
        'test_files': [a['file'] for a in analyses],
        'naming_pattern': detect_naming_pattern(test_files),
    }


def find_test_files(directory):
    """Find all test files in directory."""
    test_files = []
    path = Path(directory)

    if path.is_file():
        return [path] if is_test_file(path) else []

    # Search for test files
    for pattern in ['test_*.py', '*_test.py']:
        test_files.extend(path.rglob(pattern))

    return sorted(test_files)


def is_test_file(file_path):
    """Check if a file is a test file."""
    name = file_path.name
    return (
        name.startswith('test_') or
        name.endswith('_test.py') or
        'test' in name.lower()
    ) and name.endswith('.py')


def detect_naming_pattern(test_files):
    """Detect the naming pattern used for test files."""
    if not test_files:
        return 'unknown'

    prefix_count = sum(1 for f in test_files if Path(f).name.startswith('test_'))
    suffix_count = sum(1 for f in test_files if Path(f).name.endswith('_test.py'))

    if prefix_count > suffix_count:
        return 'prefix (test_*.py)'
    elif suffix_count > prefix_count:
        return 'suffix (*_test.py)'
    else:
        return 'mixed'


def print_analysis(analysis):
    """Pretty print the analysis results."""
    if 'error' in analysis:
        print(f"Error: {analysis['error']}")
        print(f"Searched for: {', '.join(analysis['searched_patterns'])}")
        return

    print("=" * 60)
    print("PYTHON TEST ANALYSIS")
    print("=" * 60)
    print(f"\nTest Files Found: {analysis['test_files_count']}")
    print(f"Total Tests: {analysis['total_tests']}")
    print(f"Primary Framework: {analysis['primary_framework']}")
    print(f"Naming Pattern: {analysis['naming_pattern']}")
    print(f"\nMocking Usage: {analysis['mocking_usage']}")
    print(f"Parametrize Usage: {analysis['parametrize_usage']}")

    if analysis['common_fixtures']:
        print(f"\nCommon Fixtures ({len(analysis['common_fixtures'])}):")
        for fixture in sorted(analysis['common_fixtures'])[:10]:
            print(f"  - {fixture}")
        if len(analysis['common_fixtures']) > 10:
            print(f"  ... and {len(analysis['common_fixtures']) - 10} more")

    print(f"\nTest Files:")
    for file_path in analysis['test_files'][:5]:
        print(f"  - {file_path}")
    if len(analysis['test_files']) > 5:
        print(f"  ... and {len(analysis['test_files']) - 5} more")

    print("=" * 60)


def main():
    """Main entry point."""
    if len(sys.argv) < 2:
        print("Usage: python test_analyzer.py <test_directory_or_file>")
        print("\nExample:")
        print("  python test_analyzer.py tests/")
        print("  python test_analyzer.py tests/test_user_service.py")
        sys.exit(1)

    target = sys.argv[1]

    if not os.path.exists(target):
        print(f"Error: Path '{target}' does not exist")
        sys.exit(1)

    if os.path.isfile(target):
        # Analyze single file
        analysis = analyze_test_file(target)
        print("=" * 60)
        print(f"ANALYSIS: {analysis['file']}")
        print("=" * 60)
        print(f"Framework: {analysis['framework']}")
        print(f"Test Count: {analysis['test_count']}")
        print(f"Test Classes: {', '.join(analysis['test_classes']) or 'None'}")
        print(f"Fixtures: {', '.join(analysis['fixtures']) or 'None'}")
        print(f"Uses Mocking: {analysis['uses_mocking']}")
        print(f"Uses Parametrize: {analysis['uses_parametrize']}")
    else:
        # Analyze directory
        analysis = analyze_directory(target)
        print_analysis(analysis)


if __name__ == '__main__':
    main()
